ton_lib/src/meta_loader/ipfs_loader.rs

```diff
@@ -0,0 +1,160 @@
+use reqwest::StatusCode;
+use std::fmt::Debug;
+use thiserror::Error;
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Error)] // TODO(TIAZH): Probably another file
+pub enum IpfsLoaderError {
+    #[error("Failed to load IPFS object (path: {path}, status: {status}, message: {message})")]
+    IpfsLoadObjectFailed {
+        path: String,
+        status: StatusCode,
+        message: String,
+    },
+
+    #[error("Transport error: {0}")]
+    TransportError(#[from] reqwest::Error),
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, Hash)]
+#[serde(rename_all = "snake_case")]
+pub enum IpfsConnectionType {
+    HttpGateway,
+    IpfsNode,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, Hash)]
+#[serde(rename_all = "snake_case")]
+pub struct IpfsLoaderConfig {
+    connection_type: IpfsConnectionType,
+    base_url: String,
+}
+
+impl IpfsLoaderConfig {
+    pub fn http_gateway(url: &str) -> IpfsLoaderConfig {
+        IpfsLoaderConfig {
+            connection_type: IpfsConnectionType::HttpGateway,
+            base_url: url.to_string(),
+        }
+    }
+
+    pub fn ipfs_node(url: &str) -> IpfsLoaderConfig {
+        IpfsLoaderConfig {
+            connection_type: IpfsConnectionType::IpfsNode,
+            base_url: url.to_string(),
+        }
+    }
+}
+
+impl Default for IpfsLoaderConfig {
+    fn default() -> Self { Self::http_gateway("https://cloudflare-ipfs.com/ipfs/") }
+}
+
+#[derive(Default)]
+pub struct IpfsLoaderBuilder {
+    ipfs_loader_config: Option<IpfsLoaderConfig>,
+    client: Option<reqwest::Client>,
+}
+
+impl IpfsLoaderBuilder {
+    pub fn with_config(mut self, config: IpfsLoaderConfig) -> Self {
+        self.ipfs_loader_config = Some(config);
+        self
+    }
+
+    pub fn with_client(mut self, client: reqwest::Client) -> Self {
+        self.client = Some(client);
+        self
+    }
+
+    pub fn build(self) -> IpfsLoader {
+        let config = self.ipfs_loader_config.unwrap_or_default();
+        IpfsLoader {
+            connection_type: config.connection_type,
+            base_url: config.base_url,
+            client: self.client.unwrap_or(reqwest::Client::builder().build().unwrap()),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct IpfsLoader {
+    connection_type: IpfsConnectionType,
+    base_url: String,
+    client: reqwest::Client,
+}
+
+impl Default for IpfsLoader {
+    fn default() -> Self {
+        let cfg = IpfsLoaderConfig::default();
+        Self {
+            connection_type: cfg.connection_type,
+            base_url: cfg.base_url,
+            client: Default::default(),
+        }
+    }
+}
+
+impl IpfsLoader {
+    pub fn builder() -> IpfsLoaderBuilder { IpfsLoaderBuilder::default() }
+    pub fn new() -> Self { Self::builder().build() }
```

Comment:
Remove this function, Default is too much already (but we've discussed it, it's fine)

We shouldn't expose many ways to do the same thing